src/BinaryPack/Extensions/System/TypeExtensions.cs

```diff
@@ -56,17 +55,66 @@ orderby member.Name
                       (mode.HasFlag(SerializationMode.NonPublicMembers) || member.IsPublic() ||
                        member.IsDefined(typeof(SerializableMemberAttribute))) &&
                       (member is FieldInfo fieldInfo && !fieldInfo.IsInitOnly && !fieldInfo.IsLiteral ||
-                       member is PropertyInfo propertyInfo && propertyInfo.CanRead && propertyInfo.CanWrite
+                       member is PropertyInfo propertyInfo && propertyInfo.CanRead
                        && propertyInfo.GetIndexParameters().Length == 0)
-                orderby member.Name
                 select member
-            }).ToArray();
+            }).SortMemberAsConstructorParameters(type)
+              .ToArray();
 
             return members;
         }
 
+        
+        /// Finding a constructor where all serialized members have a parameter with same type and same name (ignoring case)
+        private static IEnumerable<MemberInfo> SortMemberAsConstructorParameters(this IEnumerable<MemberInfo> members, Type type)
+        {
+            // Checking for constructor with no parameter and even any constructors first
+            if (type.GetConstructor(Type.EmptyTypes) != null || type.GetConstructors().Length == 0)
+            {
+                // Filter out properties without setter
+                return members.Where(member => member is not PropertyInfo propertyInfo || propertyInfo.CanWrite).OrderBy(member => member.Name);
+            }
+            
+            // Caching to prevent multiple iterations.
+            IEnumerable<MemberInfo> memberInfos = members as MemberInfo[] ?? members.ToArray();
+
+            List<MemberInfo> orderedMembers = new(memberInfos.Count());
+            
+            foreach (ConstructorInfo constructor in type.GetConstructors())
+            {
+                if (constructor.GetParameters().Length != memberInfos.Count())
```

Comment:
This rejects constructors that should (in theory) be valid because `memberInfos` also contains inherited members. The base `Packet` class has two members that never show up in constructors for derived types, so the program will treat those as invalid. Ideally, it should be able to recognize partial constructors and fill in the remaining properties using setters.

Alternatively, these properties are normally set to default values in the constructors, so another approach would be to create an attribute that causes the code finding parameter constructors to ignore certain properties.

Reply:
Shouldn't the members you mentioned in `Packet.cs` (which shouldn't be serialized in the first place) marked with `IgnoredMemberAttribute`?

The other problem is with types that set inherited members in the constructor without a corresponding parameter. The only option I see to solve this is to have an attribute which forces to use a given constructor. Are you fine with this approach?